pkg/controller/infrastructure/configvalidator.go

```diff
@@ -0,0 +1,102 @@
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/helper"
+	gcpclient "github.com/gardener/gardener-extension-provider-gcp/pkg/gcp/client"
+	"github.com/gardener/gardener/extensions/pkg/controller/infrastructure"
+
+	"github.com/gardener/gardener/extensions/pkg/controller/common"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// configValidator implements ConfigValidator for GCP infrastructure resources.
+type configValidator struct {
+	common.ClientContext
+	gcpClientFactory gcpclient.Factory
+	logger           logr.Logger
+}
+
+// NewConfigValidator creates a new ConfigValidator.
+func NewConfigValidator(gcpClientFactory gcpclient.Factory, logger logr.Logger) infrastructure.ConfigValidator {
+	return &configValidator{
+		gcpClientFactory: gcpClientFactory,
+		logger:           logger.WithName("gcp-infrastructure-config-validator"),
+	}
+}
+
+// Validate validates the provider config of the given infrastructure resource with the cloud provider.
+func (c *configValidator) Validate(ctx context.Context, infra *extensionsv1alpha1.Infrastructure) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	logger := c.logger.WithValues("infrastructure", client.ObjectKeyFromObject(infra))
+
+	// Get provider config from the infrastructure resource
+	config, err := helper.InfrastructureConfigFromInfrastructure(infra)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(nil, err))
+		return allErrs
+	}
+
+	// Create GCP compute client
+	computeClient, err := c.gcpClientFactory.NewComputeClient(ctx, c.Client(), infra.Spec.SecretRef)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(nil, err))
+		return allErrs
+	}
+
+	// Validate infrastructure config
+	switch {
+	case config.Networks.CloudNAT != nil:
+		logger.Info("Validating infrastructure networks.cloudNAT configuration")
+		allErrs = append(allErrs, c.validateCloudNAT(ctx, computeClient, infra.Spec.Region, config.Networks.CloudNAT, field.NewPath("networks", "cloudNAT"))...)
+	}
+
+	return allErrs
+}
+
+func (c *configValidator) validateCloudNAT(ctx context.Context, computeClient gcpclient.ComputeClient, region string, cloudNAT *api.CloudNAT, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(cloudNAT.NatIPNames) == 0 {
+		return allErrs
+	}
+
+	// Get external IP addresses mapped to whether they are available or not
+	externalAddresses, err := computeClient.GetExternalAddresses(ctx, region)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, err))
+		return allErrs
+	}
+
+	// Check whether each specified NAT IP name exists and is available
+	for i, natIP := range cloudNAT.NatIPNames {
+		natIPNamePath := fldPath.Child("napIPNames").Index(i).Child("name")
+		if available, ok := externalAddresses[natIP.Name]; !ok {
```

Comment:
The check for `GetExternalAddresses` returned map is:

```
addresses[address.Name] = address.Status == "RESERVED"
```

Am I wrong then in thinking that if the value returned is _true_ then the nat ip is _unavailable_ ? If yes, that would mean we have to fix the name or the respective condition.

Reply:
`true` here means the nat ip is *available* as originally intended, see https://pkg.go.dev/google.golang.org/api@v0.54.0/compute/v1#Address:

```
	// Status: [Output Only] The status of the address, which can be one of
	// RESERVING, RESERVED, or IN_USE. An address that is RESERVING is
	// currently in the process of being reserved. A RESERVED address is
	// currently reserved and available to use. An IN_USE address is
	// currently being used by another resource and is not available.
```

Reply:
Also, I have of course tested the above implementation and verified that it works as intended, including for IP addresses that are not available.

Reply:
@stoyanr thank you for the explanation. That makes sense .